server/etcdserver/api/membership/cluster.go

```diff
@@ -484,42 +499,68 @@ func (c *RaftCluster) PromoteMember(id types.ID, shouldApplyV3 ShouldApplyV3) {
 	c.Lock()
 	defer c.Unlock()
 
-	c.members[id].RaftAttributes.IsLearner = false
-	c.updateMembershipMetric(id, true)
+	if _, ok := c.members[id]; ok {
+		// TODO: move this line into the `if c.be != nil && shouldApplyV3 {`
+		// branch when v2store is removed.
+		c.members[id].RaftAttributes.IsLearner = false
+	}
+
 	if c.v2store != nil {
 		mustUpdateMemberInStore(c.lg, c.v2store, c.members[id])
 	}
+
 	if c.be != nil && shouldApplyV3 {
+		c.updateMembershipMetric(id, true)
 		c.be.MustSaveMemberToBackend(c.members[id])
-	}
 
-	c.lg.Info(
-		"promote member",
-		zap.String("cluster-id", c.cid.String()),
-		zap.String("local-member-id", c.localID.String()),
-	)
+		c.lg.Info(
+			"promote member",
+			zap.String("cluster-id", c.cid.String()),
+			zap.String("local-member-id", c.localID.String()),
+		)
+	} else {
+		c.lg.Info(
+			"ignore already promoted member",
+			zap.String("cluster-id", c.cid.String()),
+			zap.String("local-member-id", c.localID.String()),
+		)
+	}
 }
 
 func (c *RaftCluster) UpdateRaftAttributes(id types.ID, raftAttr RaftAttributes, shouldApplyV3 ShouldApplyV3) {
 	c.Lock()
 	defer c.Unlock()
 
-	c.members[id].RaftAttributes = raftAttr
+	if _, ok := c.members[id]; ok {
+		// TODO: move this line into the `if c.be != nil && shouldApplyV3 {`
+		// branch when v2store is removed.
+		c.members[id].RaftAttributes = raftAttr
```

Comment:
Because the member data needs to be saved to both v2store and v3store, so we have to update the member info beforehand, so that both v2 and v3 save the latest data.

Once the v2store is removed, we can move it into the `if c.be != nil && shouldApplyV3 {` branch (line 544).

Reply:
Ok, I can understand that `mustUpdateMemberInStore` and `MustSaveMemberToBackend` requires a Member struct to be passed, however if you are switching apply to v3 storage, you should follow the pattern in other functions not update the `c.members[id]` if should shouldApplyV3 is false. You might need to copy Member data from c.members[id].

Reply:
> You might need to copy Member data from c.members[id].

Yes, this is a valid point at the moment. Since we will remove v2store pretty soon, so I just leave a message here. If you  insist on it (I agree with your principle "`each PR should complete on it's own and assume that it could be the last one`." ), I will make the change.

Reply:
Just updated since it's just minor effort. Please take a look at the 4th commit https://github.com/etcd-io/etcd/pull/16658/commits/353b10f5962762f18f0dc902972714165f32b921